src/coordinates/uniform_cartesian.hpp

```diff
@@ -0,0 +1,168 @@
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef COORDINATES_UNIFORM_CARTESIAN_HPP_
+#define COORDINATES_UNIFORM_CARTESIAN_HPP_
+
+#include <array>
+#include <cassert>
+
+#include "athena.hpp"
+#include "basic_types.hpp"
+
+#include <Kokkos_Macros.hpp>
+
+namespace parthenon {
+
+class UniformCartesian {
+ public:
+  UniformCartesian() = default;
+  UniformCartesian(const RegionSize &rs, ParameterInput *pin) {
+    dx_[0] = (rs.x1max - rs.x1min) / rs.nx1;
+    dx_[1] = (rs.x2max - rs.x2min) / rs.nx2;
+    dx_[2] = (rs.x3max - rs.x3min) / rs.nx3;
+    area_[0] = dx_[1] * dx_[2];
+    area_[1] = dx_[0] * dx_[2];
+    area_[2] = dx_[0] * dx_[1];
+    cell_volume_ = dx_[0] * dx_[1] * dx_[2];
+    istart_[0] = NGHOST;
+    istart_[1] = (rs.nx2 > 1 ? NGHOST : 0);
+    istart_[2] = (rs.nx3 > 1 ? NGHOST : 0);
+    xmin_[0] = rs.x1min - istart_[0] * dx_[0];
+    xmin_[1] = rs.x2min - istart_[1] * dx_[1];
+    xmin_[2] = rs.x3min - istart_[2] * dx_[2];
+  }
+  UniformCartesian(const UniformCartesian &src, int coarsen) {
+    istart_ = src.GetStartIndex();
+    dx_ = src.Dx();
+    xmin_ = src.GetXmin();
+    xmin_[0] += istart_[0] * dx_[0] * (1 - coarsen);
+    xmin_[1] += istart_[1] * dx_[1] * (1 - coarsen);
+    xmin_[2] += istart_[2] * dx_[2] * (1 - coarsen);
+    dx_[0] *= coarsen;
+    dx_[1] *= (istart_[1] > 0 ? coarsen : 1);
+    dx_[2] *= (istart_[2] > 0 ? coarsen : 1);
+    area_[0] = dx_[1] * dx_[2];
+    area_[1] = dx_[0] * dx_[2];
+    area_[2] = dx_[0] * dx_[1];
+    cell_volume_ = dx_[0] * dx_[1] * dx_[2];
+  }
+
+  template <class... Args>
+  KOKKOS_FORCEINLINE_FUNCTION Real Volume(Args... args) const {
+    return cell_volume_;
+  }
+
+  const std::array<Real, 3> &Dx() const { return dx_; }
+
+  template <class... Args>
+  KOKKOS_FORCEINLINE_FUNCTION Real Dx(const int dir, Args... args) const {
+    assert(dir > 0 && dir < 4);
+    return dx_[dir - 1];
+  }
+
+  const std::array<Real, 3> &Area() const { return area_; }
+
+  template <class... Args>
+  KOKKOS_FORCEINLINE_FUNCTION Real Area(const int dir, Args... args) const {
+    assert(dir > 0 && dir < 4);
+    return area_[dir - 1];
+  }
```

Comment:
Although I like that you chose to call this area instead of faceArea, would this cause some confusion in 2D simulations where the geometric 'area' is equivalent of the volume?  Or, is it the case that we **always** run 3D, so the distinction is not necessary?

Reply:
I don't think it's confusing.  In truth, these actually are always volumes, areas, etc. because there are always 3 dimensions defined.